serctl/serxmlrpc.3
.TH serxmlrpc 3 2006-01-17 "" "Python serxmlrpc module. "
.SH NAME
serxmlrpc \- extension of xmlrpclib module for digest authorization and
persistent connection. Mainly for use with SER (Sip Express Router) as
client side library.
.SH SYNOPSIS
.nf
.B from serctl.serxmlrpc import *
.sp
.IB \fPclass\  Transport( uri =None, \ ssl =None)
.sp
.IB \fPexception\  UnsupportedAuthScheme( scheme )
.sp
.IB \fPfunction\  ServerProxy( uri , \ ssl =None, \ encoding =None, \ verbose =0, \ allow_none =0)
.fi
.SH DESCRIPTION
Instance of \fBTransport\fP class can be used as transport instance for \fBxmlrpc.ServerProxy\fP.
.TP
.B uri
is URI string of the form http[s]://[user[:password]@]host[:port][/path]. It is
used for making connection to server. When omited or set to None default
values are used: user='anonymous', password='', host='localhost', port=5060.
User and password are used for digest authorization.
.TP
.B ssl
is tuple of strings (ssl_key_file, ssl_cert_file), where ssl_key_file and ssl_cert_file 
are names of private PEM formated key file and PEM formated certificate
chain file. Default is None, it mean no ssl authentication is used.
.PP
Exception
.B UnsupportedAuthScheme
is raised when server require unsupported authorization scheme.
Only digest authorization scheme supported, now.
.PP
Function
.B ServerProxy
is wrapper around ServerProxy class from xmlrpclib module. It instanties xmlrpclib.ServerProxy
with transport parameter set to new \fBTransport\fP instance. Uri and ssl
parameters are the same as described above for \fBTransport\fP. All other
parameters are directly passed to xmlrpclib.ServerProxy. See documentation
for xmlrpclib. 
.SH RETURN VALUE
ServerProxy function return xmlrpclib.ServerProxy instance.
.SH PYTHON VERSION
Module serxmlrpc is for python 2.2+. 
.SH NOTE
Serxmlrpc module can be used separately. It's not dependent on serctl package.


sip_router/NEWS
Release notes for SIP Express Router (ser)
***********************************************

$Id: NEWS,v 1.49 2005/12/12 23:47:56 andrei Exp $

0.10.99-dev changes


new archs:


modules:
 - dispatcher - added hashing after request uri and to uri
              - added a new flag parameter which can be used (for now) to
                select only the username or the username, host and port when
                hashing after an uri (to, from or request uri)
              - improved uri hashing (password is ignored, port is used only
                if != 5060 or 5061)
 - tm        - migrated to the new timers (tm timers completely rewritten)
             - improved speed and less memory usage
             - much more precise reptransmissions timing
             - params: - retr_timer1p1, retr_timer1p2, retr_timer1p3 removed
                         and replaced by retr_timer1 and retr_timer2
                       - all timer values are now expressed in milliseconds
                         (they were in seconds before).
                         Affected params: fr_timer, fr_inv_timer, wt_timer,
                         delete_timer, retr_timer1, retr_timer2
                       - retr_timer1 (first retransmission) changed to 500 ms
                       - delete_timer changed to 200 ms
                       - unix_tx_timeout expressed now in milliseconds; default
                         value changed to 500 ms
             - functions:
                       - new t_set_fr(timeout_fr_inv, timeout_fr) -- allows
                         changing the transaction timer from script, even if
                         the transaction was already created (see tm docs for
                         more).
            - t_relay will not stop script execution anymore in case of
              send error
 - textops  - search() can be used in the onsend_route where it will search
              on the "new" message (after applying all script changes, adding
              Vias a.s.o) and not on the original message
 
core:
 - added return [val] which returns from a route. if no value is specified, or
   a route reaches its end without executing a return statement, it returns 1.
   If return is used in the top level route is equivalent with exit [val].
 - drop /exit [n] now will end the script execution
   exit n will exit with code n (usefull in onreply/onsend routes where
   if script code !=0 a reply is generated/the message is sent or to force
   script errors)
 - added $? which can be used to check the return code of the last executed
   route{} (e.g. route(1); if ($?==1){ /* ... */}else if ($?==2) ... )
 - onsend_route added: special route executed before a request is sent.
                       Only a limited number of commands are allowed (drop, if
                       + all the checks, msg flag manipulations, send(), log(),
                       textops::search()). In this route the final destination
                       of the message is available an can be checked (with
                       snd_ip, snd_port, to_ip, to_port, snd_proto, snd_af).
                       This route is executed only when forwarding requests.
                       It's not executed for replies, retransmissions, or
                       locally generated messages (e.g. via fifo uac).
      short example:
            onsend_route{  if(to_ip==1.2.3.4 && !isflagset(12)){
                              log(1, "message blocked\n");
                              drop;
                           }
                         }
 - onsend_route specific checks:
     - snd_ip, snd_port - behave like src_ip/src_port, but contain the
       ip/port ser will use to send the message
     - to_ip, to_port - like above, but contain the ip/port the message will
       be sent to (not to be confused with dst_ip/dst-port, which are the
       destination of the original message: ser's ip and port on which the
       message was received)
     - snd_proto, snd_af - behave like proto/af but contain the 
       protocol/address family that ser will use to send the message
     - msg:len - when used in an onsend_route, msg:len will contain the length
       of the message on the wire (after all the changes in the script are
       applied, Vias are added a.s.o) and not the lentgh of the original 
       message
 - timer: - improved performance/precision, new api, see doc/timers.txt 
 - tcp: - improved  performance (io event handling), using OS specific
           optimizations
        - 1024 connections limit removed (see tcp_max_connections)
-  resolver: - timeouts, retries a.s.o can no be set from ser.cfg
             (see below dns_* and man resolv.conf(6)).
             The maximum time a dns request can take (before failing) is:
              (dns_retr_time*dns_retr_no)*(search_list_domains)
             If dns_try_ipv6 is yes, mutliply it again by 2.
             The fastest possible dns config (max 1s):
                dns_try_ipv6=no
                dns_retr_time=1
                dns_retr_no=1
                dns_use_search_list=no
- default on reply route added: onreply_route {.. } will add a default 
    onreply route that will be executed for any reply (usefull to catch
    replies without using tm)
- branch_routes added (tm triggered), only a very limited number of commands
   are available (see tm docs)
- avps directly accessible from script with %avp_name (variable style)
new config variables:
   tos = number  - ip type of service (TOS) value
   dns_try_ipv6 = yes/no - if yes and a dns lookup fails, it will retry it
      for ipv6 (AAAA record). Default: yes
   dns_retr_time = time - time in s before retrying a dns request.
      Default: system specific, depends also on the/etc/resolv.conf content
      (usually 5s).
   dns_retr_no = no. - number of dns retransmissions before giving up.
      Default: see above (usually 4)
   dns_server_no = no. - how many dns servers from the ones defined in 

serctl/serctl/serxmlrpc.py
#!/usr/bin/env python
# -*- encoding: UTF-8 -*-
#
# $Id: serxmlrpc.py,v 1.2 2006/01/18 11:01:44 hallik Exp $
#
# Copyright (C) 2005 iptelorg GmbH
#
# This is part of SER (SIP Express Router), a free SIP server project. 
# You can redistribute it and/or modify it under the terms of GNU General
# Public License as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#

from urllib  import unquote
from urllib2 import parse_http_list, parse_keqv_list
import httplib, xmlrpclib, re, md5

# indexes to parsed uri sequence
SCHEME = 0
USER   = 1
PASS   = 2
HOST   = 3
PORT   = 4

_EXP = re.compile('(^[^:]*)://([^/]*)/?.*$')

def _split_uri(uri):
	exp      = _EXP.match(uri)
	if not exp:
		return ('http', None, '', 'localhost', '5060')
	scheme   = exp.group(1)
	uphp     = exp.group(2)
	if not uphp:
		return (scheme, None, '', 'localhost', '5060')
	uphp = uphp.split('@')
	if len(uphp) < 2:
		user = None
		passwd = ''
		hp = uphp[0].split(':')
		if len(hp) < 2:
			port = '5060'
		else:
			port = hp[1]
		host = hp[0]
	else:
		up = uphp[0].split(':')
		if len(up) < 2:
			passwd = ''
		else:
			passwd = up[1]
		user = up[0]
		hp = uphp[1].split(':')
		if len(hp) < 2:
			port = '5060'
		else:
			port = hp[1]
		host = hp[0]
	return (scheme, user, passwd, host, port)

def _unquote(i):
	if not i:
		return i
	return unquote(i)

def _parse_uri(uri):
	return [ _unquote(i) for i in _split_uri(uri) ]

class UnsupportedAuthScheme(httplib.HTTPException):
	def __init__(self, scheme):
		self.args = (scheme, )
		self.scheme = scheme

class Transport:
	HEADERS = {"Content-type": "application/x-www-form-urlencoded",
	           "User-Agent"  : "serctl/1.0" }

	def __init__(self, uri=None, ssl=None):
		puri = _parse_uri(uri)
		puri[PORT] = int(puri[PORT])
		if puri[USER]:
			self.user = puri[USER]
			self.password = puri[PASS]
		else:
			self.user = 'anonymous'
			self.password = ''
		self.auth = None

		if puri[SCHEME].lower() == 'https':
			if ssl:
				self.conn = httplib.HTTPSConnection( \
				  puri[HOST], puri[PORT], ssl[0], ssl[1])
			else:
				self.conn = httplib.HTTPSConnection( \
				  puri[HOST], puri[PORT])
		else:
			self.conn = httplib.HTTPConnection(puri[HOST], \
			  puri[PORT])

	def _http_request(self, uripath, body, host):
		headers = self.HEADERS.copy()
		headers['Host'] = host
		if self.auth:
			headers['Authorization'] = self.auth
		self.conn.request("POST", uripath, body, headers)

	def _generate_auth_header(self, scheme, uripath, realm, nonce):
		# variable scheme is silently ignored. It is intended for 
		# multiple authentication schemes - only MD5 scheme supported,
		# now. Value of this variable should be equivalent to 
		# lowercased string of HTTP auth scheme name.
		a1 = "%s:%s:%s" % (self.user, realm, self.password)
		a2 = "POST:" + uripath
		h1 = md5.new(a1).hexdigest()
		h2 = md5.new(a2).hexdigest()
		a3 = "%s:%s:%s" % (h1, nonce, h2)
		h3 = md5.new(a3).hexdigest()
		auth = 'Digest username="%s", realm="%s", nonce="%s", ' \
		       'uri="%s", response="%s"' % (self.user, realm, nonce, \
		        uripath, h3)
		return auth

	def _parse_auth_header(self, authhdr):
		l = authhdr.split(' ', 1)
		if len(l) != 2:
			return None, {}
		method = l[0].lower()
		challenge = parse_keqv_list(parse_http_list(l[1]))
		return method, challenge

	def request(self, host, uripath, body, verbose=0):

		self._http_request(uripath, body, host)
		response = self.conn.getresponse()

 		if response.status == 401 or response.status == 407:
			response.read()
			if response.status == 401:
				hdrname = 'WWW-Authenticate'
			else:
				hdrname = 'Proxy-Authenticate'
			authhdr = response.getheader(hdrname)
			scheme, challenge = self._parse_auth_header(authhdr)
			if scheme != 'digest':
				raise UnsupportedAuthScheme (scheme,)

			self.auth = self._generate_auth_header(self, uripath, \
			  challenge['realm'], challenge['nonce'])

			self._http_request(uripath, body, host)
			response = self.conn.getresponse()

		if response.status != 200:
			raise xmlrpclib.ProtocolError(host + uripath, \
			response.status, response.reason, response.msg)

		data = response.read()

		parser, unmarshaller = xmlrpclib.getparser()
		parser.feed(data)
		parser.close()
		return unmarshaller.close()

def ServerProxy(uri, ssl=None, encoding=None, verbose=0, allow_none=0):
	transport = Transport(uri, ssl)
	return xmlrpclib.ServerProxy(uri, transport, encoding, verbose, \
	  allow_none)
